### PR TITLE
feat: initial relay integration support

### PR DESCRIPTION
**Pull request by bellini666**
TODO:
- [x] Support for connections
- [x] Support for `node()`

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)

File: tests/relay/test_node.py
Comment (mattalbr) [inline tests/relay/test_node.py:34]: Do you think there's any way to create this type by default in the library through primary key abstraction? Doesn't have to be in this PR, but I really would love us to get to zero boilerplate

**Review by mattalbr — COMMENTED** [review]
I'm still reviewing, but this is definitely over my head so far. Trying to focus on the API since I'm still wrapping my head around the implementation. Going to relay on @erikwrede and @patrick91 heavily :)

**Comment by erikwrede** [comment]
hey everyone, I still need some time to fully review this. Will get back when I have done some more testing

**Comment by bellini666** [comment]
Hey @erikwrede , @mattalbr 

I just got back from djangocon and I'll get back to this PR during the weekend now.

One thing I was wondering here is regarding the fact that you need to have a session to query sqlalchemy objects.

Because of that I had to add a `sessionmaker` attribute to the fields, and I notice that the data loaders require one as well.

I was thinking about a way to avoid that and I have a proposal. Let me know what you think and I can implement that in this PR:

The idea is to use [contextvars](https://docs.python.org/3/library/contextvars.html) to provide a session that will be used during the execution of a query. Not only the relay fields I'm defining here can use it, but the dataloaders can also use it instead of requiring a `bind` argument.

Something like:

```python
current_session = contextvars.ContextVar("current-session")

class SessionProviderExtension(SchemaExtension):
    def __init__(self, sessionmaker):
        self.sessionmaker = sessionmaker

    def on_execute(self):
        token = current_session.set(self.sessionmaker())
        try:
            yield
        finally:
            current_session.reset(token)
```

Of course we can also make the current_session be started lazely (create it when first accessing it)

What do you think?

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/field.py:100]: This is to support limit/offset from `ListConnection`

I just implemented a custom connection using the sqlkeyset pagination integration

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/field.py]: This is specifically for nested connections, which I didn't even resolve yet strawberry django due to it not being very easy to solve when you are paginating the inner portions 

I added a `TODO` note in here for now

File: tests/relay/test_node.py
Comment (bellini666) [reply tests/relay/test_node.py:34]: Probably there is a way, indeed! It goes beyond the scope of this PR, but it can be done in the future

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [inline src/strawberry_sqlalchemy_mapper/field.py:552]: nit: I think this documentation would be a bit clearer if this used `class Query`

File: tests/relay/test_node.py
Comment (mattalbr) [inline tests/relay/test_node.py:66]: Trying to understand how this relay API works. Having this returned in `id` seems like it would obfuscate `id` to the caller, which for a lot of situations is totally fine, but what if we wanted to do this without having an opaque `id`? Is there a way to put the relay ID in a different response field, but still tell the mapper that it should use `id` for mapping?

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [inline src/strawberry_sqlalchemy_mapper/field.py:332]: I almost want to log a warning here because this will be so bad for performance. But, I think we'd only want to log it once per query, rather than per resolver, which maybe we don't have an easy way to do. So maybe no action, just thinking out loud.

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [inline src/strawberry_sqlalchemy_mapper/field.py:494]: nit: should be strawberry_sqlalchemy_mapper not django :)

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [inline src/strawberry_sqlalchemy_mapper/field.py:684]: I think `first` and `last` should be `int`s, right?

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [inline src/strawberry_sqlalchemy_mapper/field.py:674]: How would someone do something like this were they extend the resolver to implement custom filtering without pulling all the data in from the database? Feels like either an abstraction is missing that would allow this to modify the query instead of returning results, or it would need to have access to the pagination params

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [inline src/strawberry_sqlalchemy_mapper/field.py:89]: This is at field definition time, correct? Keyset I believe should come not at definition time but at query time. The Keyset gets passed to the caller as the pagination token (well technically the Marker gets passed, which includes the Keyset), which gets passed back when fetching the next page.

**Review by mattalbr — COMMENTED** [review]
Just a few API clarifications and concerns, but this is AWESOME and a huge step in the right direction for this library.

Lots of things we can improve eventually, but I love the direction this is going. Thank you so much Bellini!

Patrick and Erik, I'm relying on you two to review the implementation since it's way out of my expertise, and I'm strictly keeping a user-focused lens while reviewing.

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (erikwrede) [inline src/strawberry_sqlalchemy_mapper/relay.py:258]: DjangoOptimizer mention is probably an artefact from the django implementation, right?

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (erikwrede) [inline src/strawberry_sqlalchemy_mapper/relay.py:391]: any reason to not internally re-use `resolve_model_nodes` here?

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (erikwrede) [inline src/strawberry_sqlalchemy_mapper/relay.py:386]: Again, we should be clear about where async is supported. Not a blocker for first version, we can add async everywhere later.

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (erikwrede) [inline src/strawberry_sqlalchemy_mapper/relay.py:394]: Can't we handle this via `str | tuple[str]`? Parsing a string everytime seems slow and hard to understand without knowing the source. Also easier to not miss the standard for users providing custom implementations

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (erikwrede) [inline src/strawberry_sqlalchemy_mapper/relay.py:424]: Let's find a way to cache this on the type later, no need to re-compute on every call

File: tests/relay/test_node.py
Comment (erikwrede) [reply tests/relay/test_node.py:34]: There definitely is a way, but the zero boilerplate should be optional. Explicitly selecting what to include in the schema on a type level has advantages too, so I'm all for supporting both ways.

File: tests/relay/test_node.py
Comment (erikwrede) [reply tests/relay/test_node.py:66]: Since this is the resolved query to the frontend, the frontend clients should only know the real relay ID. In your mutations you can use `relay.from_base64` and then the id to node function to get the corresponding DB entity. Where else would you need the raw ID where it is difficult to access by this paradigm?

Not sure I got the question right, so please feel free to specify.

**Review by erikwrede — CHANGES_REQUESTED** [review]
Looks great!

Minor revisions attached, let's figure those points out and work on shaping the rest iteratively in future releases.

About the context vars for sessions: I'd prefer to leave things in context for now, this way we don't force async on everyone and we have a clear entry point for field extensions to modify the context and thus the session.

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/field.py:674]: By returning a `Query` object. This is actually exactly what the `default_resolver` does when you don't define one, but you could do something like:

```python
@relay.connection(relay.ListConnection[SomeType])
def some_connection(self) -> Iterable[SomeType]:
    return session.query(SomeModel)
```

In this case `session.query(SomeModel)` will only access the database when being iterated, and what you return in here will be modified by the connection later: for `ListConnection` it will slice it, producing a `limit`/`offset` pagination. For the `KeysetConnection` I created it will do its thing

An example from strawberry-django: https://github.com/bellini666/strawberry-django-demo/blob/main/demo/order/schema.py#L28

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/field.py:89]: @mattalbr not sure if I understood whast you mean, and maybe the name of the variable here is not the best one =P

This variable defines the fields that will be used for the pagination, i.e., that will be used to `order_by` the query when you are relying on the default resolver.

You can see an example in [this test](https://github.com/strawberry-graphql/strawberry-sqlalchemy/pull/65/files#diff-e466e4061d3ec646290e8d41bfb1ccfb9f3d7f36113d7c8c9196d2e1ae9de05aR509)

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/relay.py:140]: I had to use the `sqlakeyset.get_page` API instead of the `sqlakeyset.select_page` because of the way this is implemented.

sqlakeyset doesn't have a `get_page` async counterpart, only a `select_page`

TBH I don't know much of the sqlalchemy internals and neither of `sqlakeyset`, so open to try any suggestions here =P

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/relay.py:391]: I based this at strawberry-django's code, and in there I have different approaches for both functions.

Looking now you are correct, we can just reuse it :)

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/relay.py:375]: This would not work very well as the relay implementation on strawberry will do a `str()` on anything that is not already a string, meaning `(1, 2, 3)` would become `"(1, 2, 3)"`, which is harder to parse

But TBF, multiple primary keys are not very common, so anyone using them would probably know that there must be some special handling for this.

But I can/will add a comment here about this

**Comment by bellini666** [comment]
@mattalbr @erikwrede think I addressed all of the comments in this PR

Let me know if there's anything else missing to be able to merge this

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (patrick91) [reply src/strawberry_sqlalchemy_mapper/relay.py:379]: sounds good to me 😊

Comment (erikwrede) [reply src/strawberry_sqlalchemy_mapper/relay.py:379]: @bellini666 @mattalbr are you fine with merging this in and later causing a soft breaking change in the internal api by switching to tuples after we added support in strawberrry? I'd like to avoid blocking the PR for that.

**Review by erikwrede — APPROVED** [review]
lgtm from my side, pending agreement on this from you:

> @bellini666 @mattalbr are you fine with merging this in and later causing a soft breaking change in the internal api by switching to tuples after we added support in strawberrry? I'd like to avoid blocking the PR for that.

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/relay.py:379]: I'm fine with this :)

**Comment by bellini666** [comment]
> lgtm from my side, pending agreement on this from you:

Thank you! @mattalbr let me know if you want any extra changes in this

File: tests/relay/test_node.py
Comment (mattalbr) [reply tests/relay/test_node.py:66]: It's more a matter of, what if someone already has a field named ID that the client is already using, but they want to add support using this new relay functionality. It's nice to think all ids should be opaque, but that's far too opinionated for a middleware library like this to make that decision IMO.

So my question is, what does someone do if they already have a field named "id" that the client is depending on? Can they put this relay ID in a different field so as not to clobber "id"?

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [reply src/strawberry_sqlalchemy_mapper/field.py]: Amazing, thanks!

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (mattalbr) [reply src/strawberry_sqlalchemy_mapper/relay.py:379]: I'm fine with that, yeah! I'm already way too slow on this PR and gotta get out of the way.

**Review by mattalbr — APPROVED** [review]
Last of the open questions, but looks good for a first version! Thanks so much for your help!

File: tests/relay/test_node.py
Comment (bellini666) [reply tests/relay/test_node.py:66]: @mattalbr it is part of the `Node` interface and its spec to have a `id` attribute which is serialized something that can globally identify the object (i.e. without even having to tell the type).

Having said that, there's a proposal to make it easier to support custom implementations (in case you want to make it opaque in some other way, or even skip it at all) which might be useful once it gets merged: https://github.com/strawberry-graphql/strawberry/issues/3057

Another possibility is to expose the original id in another field.

But I would say that anyone wanting to use this is not only aware of the whole GlobalID thing, but also expects that it works like that.

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/field.py:89]: @mattalbr oh I see. In the way I implemented it, the ordering is defined in the field

The old link is not working anymore, here is a new one: https://github.com/strawberry-graphql/strawberry-sqlalchemy/pull/65/files#diff-e466e4061d3ec646290e8d41bfb1ccfb9f3d7f36113d7c8c9196d2e1ae9de05aR509

Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/field.py:89]: Oh, the link is not working because the file itself is collapsed.

Here is the screenshot of the implementation

Screenshot 2023-11-30 at 18 12 31

Comment (bellini666) [reply src/strawberry_sqlalchemy_mapper/field.py:89]: Having said that 

> Sorry that link isn't working for me, can you point me to the test by name? I just want to make sure that which fields will be used to order the query is defined per request. E.g. imagine we had a resolver:
> 
> ```python
> def fruits(order: list[string], info: Info):
>     return info.context.session.query(models.Fruit).order(*order).all()
> ```
> 
> The ordering, and thus the keyset, is not fixed for the resolver -- it's determined by the client.

Having said that, your example here would actually work. keyset is only used for the automatic resolver, not for manual ones. So if you define a resolver like you did it will work

File: tests/relay/test_node.py
Comment (patrick91) [reply tests/relay/test_node.py:66]: > But I would say that anyone wanting to use this is not only aware of the whole GlobalID thing, but also expects that it works like that.

I personally disagree with this one :D I think the global id is mostly useful when using relay as a client, which might not be the case most of the times 😊

I think it might be worth trying to solve 3057 in Strawberry, but I haven't had the right amount of time to think about a good solution yet

File: src/strawberry_sqlalchemy_mapper/field.py
Comment (mattalbr) [inline src/strawberry_sqlalchemy_mapper/field.py:351]: Ah! No need to add this "order" column yet since we don't have a fully fleshed API yet. I think you can just do:

```python
if after := kwargs.get("after"):
  order_by.extend(sqlakeyset.unserialize_bookmark(after).place)
```

And then specify where field.keyset is defined that it's just the default ordering, and later pages grab the order from the cursor.

**Review by mattalbr — APPROVED** [review]

**Comment by guilhermelou** [comment]
> current_session

Hey @bellini666 , have you started working on something to avoid the duplicate session setup?

**Pull request merged**
